gnucash/gnome-utils/gnc-report-combo.c

```diff
@@ -0,0 +1,595 @@
+ *  should release any memory not previously released by the destroy
+ *  function (i.e. the private data structure), then chain up to the
+ *  parent's finalize function.
+ *
+ *  @param object The object being finalized.
+ *
+ *  @internal
+ */
+static void
+gnc_report_combo_finalize (GObject *object)
+{
+    GncReportComboPrivate *priv;
+    GncReportCombo *grc;
+
+    g_return_if_fail (object != NULL);
+    g_return_if_fail (GNC_IS_REPORT_COMBO(object));
+
+    grc = GNC_REPORT_COMBO(object);
+    priv = GET_PRIVATE(grc);
+
+    if (priv->active_report_guid)
+        g_free (priv->active_report_guid);
+
+    if (priv->active_report_name)
+        g_free (priv->active_report_name);
+
+    G_OBJECT_CLASS (gnc_report_combo_parent_class)->finalize (object);
+}
+
+static gboolean
+select_active (GncReportCombo *grc)
+{
+    GncReportComboPrivate *priv = GET_PRIVATE(grc);
+    GtkTreeModel *model = gtk_combo_box_get_model (GTK_COMBO_BOX(priv->combo));
+    GtkTreeIter iter;
+    gboolean valid_iter = gtk_tree_model_get_iter_first (model, &iter);
+    gboolean found = FALSE;
+
+    while (valid_iter)
+    {
+        gchar *guid;
+        gtk_tree_model_get (model, &iter, RC_GUID, &guid, -1);
+
+        if (g_strcmp0 (priv->active_report_guid, guid) == 0)
+        {
+            gtk_combo_box_set_active_iter (GTK_COMBO_BOX(priv->combo), &iter);
+            g_free (guid);
+            found = TRUE;
+            break;
+        }
+        g_free (guid);
+        valid_iter = gtk_tree_model_iter_next (model, &iter);
+    }
+
+    if (!found)
+    {
+        GtkTreeIter iter;
+        QofBook *book = gnc_get_current_book ();
+        gchar *tmp;
+
+        if (priv->active_report_name)
+            tmp = g_strdup (priv->active_report_name);
+        else
+            tmp = g_strdup (_("Selected Report is Missing"));
+
+        gtk_list_store_prepend (GTK_LIST_STORE(model), &iter);
+        gtk_list_store_set (GTK_LIST_STORE(model), &iter,
+                            RC_NAME, tmp,
+                            RC_GUID, priv->active_report_guid,
+                            RC_MISSING, TRUE,
+                            -1);
+
+        g_free (tmp);
+        gtk_combo_box_set_active_iter (GTK_COMBO_BOX(priv->combo), &iter);
+    }
+    return found;
+}
+
+/********************************************************************
+ * update_invoice_list
+ *
+ * this procedure does the real work of displaying a sorted list of
+ * available invoice reports
+ ********************************************************************/
+static void
+update_report_list (GncReportCombo *grc)
+{
+    GncReportComboPrivate *priv = GET_PRIVATE(grc);
+    GtkTreeModel *model = gtk_combo_box_get_model (GTK_COMBO_BOX(priv->combo));
+
+    SCM template_menu_name = scm_c_eval_string ("gnc:report-template-menu-name/report-guid");
+    SCM get_rpt_guids = scm_c_eval_string (priv->rpt_guids);
+    SCM reportlist;
+    SCM rpt_guids;
```

Comment:
MVC separation failure. This is a brand-new control so let's do it right. Keep the model/controller code in gnome-business-utils.c as `GList *gnc_invoice_report_list()`; the list can be of `struct { char* report_guid; char* report_name}`.

Reply:
I was thinking that the scheme was part of as it was only to do with reports but have changed it.